### PR TITLE
Allow configurable GeoIP database path #33

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

**Review by rymsha — APPROVED** [review]

**Review by alansemenov — APPROVED** [review]

File: build.gradle [hunk 1]
```diff
@@ -6,9 +6,9 @@ plugins {
 }
 
 dependencies {
-    compileOnly "com.enonic.xp:core-api:${xpVersion}"
-    compileOnly "com.enonic.xp:portal-api:${xpVersion}"
-    compile "com.maxmind.db:maxmind-db:2.1.0"
+    implementation "com.enonic.xp:core-api:${xpVersion}"
```
Comment (rymsha) [inline build.gradle:9]: compileOnly for api is not that bad.

File: README.md [hunk 2]
```diff
@@ -10,6 +10,13 @@ The MaxMind **[GeoLite2 City](http://geolite.maxmind.com/download/geoip/database
 The location data will become less accurate over time if your database file is not updated frequently. MaxMind releases a new version of their database on the first Tuesday of each month. 
 Make sure to get the binary, gzipped version of the GeoLite2 City database, not the CSV version. 
 
+Starting from version 2.1.0 you can customize a location of the database file. In order to do that you should add the config `com.enonic.lib.geoip.database.filepath` option into the app config file, as in the example below:
+
+.com.enonic.app.appname.cfg
+```properties
+com.enonic.lib.geoip.database.filepath=/absolute/path/to/database.mmdb
```
Comment (rymsha) [inline README.md:17]:
```suggestion
com.enonic.lib.geoip.database.filepath=/path/to/database.mmdb
```

File: src/main/java/com/enonic/geoip/DbReader.java [hunk 3]
```diff
@@ -22,27 +20,34 @@
...
-        //Reader r = new Reader( db, FileMode.MEMORY_MAPPED, NoCache.getInstance() );
-        Reader r = new Reader( db, FileMode.MEMORY_MAPPED, new CHMCache() );
+    public JsonNode getLocationDataFromFile()
+        throws IOException
+    {
+        final File database =
+            new File( Objects.requireNonNullElse( databaseFilePath, System.getenv( "XP_HOME" ) + "/config/GeoLite2-City.mmdb" ) );
 
-        InetAddress ipa = InetAddress.getByName( this.ip != null ? this.ip : this.remoteAddr );
+        try (Reader r = new Reader( database, FileMode.MEMORY_MAPPED, new CHMCache() ))
```
Comment (rymsha) [inline src/main/java/com/enonic/geoip/DbReader.java:35]: Actually it looks extremely inefficient to read the database for every call

**Review by rymsha — CHANGES_REQUESTED** [review]

File: src/main/java/com/enonic/geoip/DbReader.java (hunk 3 above)
Comment (rymsha) [reply src/main/java/com/enonic/geoip/DbReader.java:35]: Maybe re-read only if file timestamp changes?

**Review by rymsha — APPROVED** [review]

**Pull request merged**